python/tempo/interpol.py

```diff
@@ -0,0 +1,322 @@
+        :ts_col int - timeseries column (must exist in the input dataframe)
+        :target_col int - column to perform interpolation (must exist in the input dataframe)
+        """
+        window_ff = (
+            Window.partitionBy(*partition_cols)
+            .orderBy(ts_col)
+            .rowsBetween(-sys.maxsize, 0)
+        )
+        window_bf = (
+            Window.partitionBy(*partition_cols)
+            .orderBy(ts_col)
+            .rowsBetween(0, sys.maxsize)
+        )
+
+        read_last = last(df[target_col], ignorenulls=True).over(window_ff)
+        read_next = first(df[target_col], ignorenulls=True).over(window_bf)
+        readtime_last = last(df["surrogate_ts"], ignorenulls=True).over(window_ff)
+        readtime_next = first(df["surrogate_ts"], ignorenulls=True).over(window_bf)
+
+        filled_series: DataFrame = (
+            df.withColumn("readvalue_ff", read_last)
+            .withColumn("readvalue_bf", read_next)
+            .withColumn("readtime_ff", readtime_last)
+            .withColumn("readtime_bf", readtime_next)
+            .filter(col("readtime_bf").isNotNull())
+        )
+
+        return filled_series
+
+    def __interpolate_column(
+        self,
+        epoch_series: DataFrame,
+        partition_cols: List[str],
+        ts_col: str,
+        target_col: str,
+        fill: str,
+    ) -> DataFrame:
+        """
+        Apply interpolation to column.
+
+        :param epoch_series  - input DataFrame in epoch seconds
+        :param ts_col   - timestamp column name
+        :param target_col   - numeric column to interpolate
+        :param fill   - interpolation function to fill missing values
+        """
+        # Generate Fill for Dataset
+        filled_series: DataFrame = self.__fill_series(
+            epoch_series, partition_cols, ts_col, target_col
+        )
+
+        output_df: DataFrame = filled_series
+
+        # Handle zero fill
+        if fill == "zero":
+            output_df = filled_series.withColumn(
+                target_col,
+                when(col(target_col).isNotNull(), col(target_col)).otherwise(lit(0)),
+            )
+
+        # Handle forward fill
+        if fill == "forward":
+            output_df = filled_series.withColumn(target_col, col("readvalue_ff"))
+
+        # Handle backwards fill
+        if fill == "back":
+            output_df = filled_series.withColumn(target_col, col("readvalue_bf"))
+
+        # Handle linear fill
+        if fill == "linear":
+            output_df = output_df.withColumn(
+                target_col,
+                self.linear_udf(
+                    ts_col,
+                    "readtime_ff",
+                    "readtime_bf",
+                    "readvalue_ff",
+                    "readvalue_bf",
+                    target_col,
+                ),
+            )
+
+        return output_df
+
+    def interpolate(
```

Comment:
@guanjieshen , I'm wondering if there is a way to extend TSDF so that this operation can be applied in the same way that 'resample' is applied. This is not required but having a consistent way to call functions on TSDF makes things simpler. Something like tsdf.interpolate as opposed to Interpolate.interpolate(tsdf)

Reply:
Yup, can definitely make that change!